rllib/utils/exploration/noveld.py

```diff
@@ -0,0 +1,463 @@
+from gym.spaces import Discrete, MultiDiscrete, Space
+import logging
+import numpy as np
+from typing import Optional, Union
+
+from ray.rllib.env.base_env import BaseEnv
+from ray.rllib.models.action_dist import ActionDistribution
+from ray.rllib.models.catalog import ModelCatalog
+from ray.rllib.models.modelv2 import ModelV2
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils import NullContextManager
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.exploration import Exploration
+from ray.rllib.utils.framework import try_import_tf, try_import_torch
+from ray.rllib.utils.from_config import from_config
+from ray.rllib.utils.tf_utils import get_placeholder
+from ray.rllib.utils.typing import FromConfigSpec, ModelConfigDict, TensorType
+
+logger = logging.getLogger(__name__)
+
+tf1, tf, tfv = try_import_tf()
+torch, nn = try_import_torch()
+F = None
+if nn is not None:
+    F = nn.functional
+
+
+class NovelD(Exploration):
+    """Implements NovelD exploration criterion.
+
+    Implementation of:
+    [1] NovelD: A simple yet Effective Exploration Criterion.
+    Zhang, Xu, Wang, Wu, Kreutzer, Gonzales & Tian (2021).
+    NeurIPS Proceedings 2021
+
+    which is based on Random Network Distillation proposed in
+    [2] Exploration By Random Network Distillation.
+    Burda, Edwards, Storkey & Klimov (2018).
+    7th International Conference on Learning Representations (ICLR 2019)
+
+    Estimates the novelty of a state by a distilled network approach.
+    The states novelty thereby increases at the boundary between explored
+    and unexplored states. It compares the prior state novelty with the
+    actual state novelty.
+
+    The novelty difference between the prior and actual state is considered
+    as intrinsic reward and added to the environment's extrinsic reward
+    for policy optimization.
+
+    Novelty is actual a very general approach and any novelty measure could be
+    used. Here the distillation error is used.
+    """
+
+    def __init__(
+        self,
+        action_space: Space,
+        *,
+        framework: str,
+        model: ModelV2,
+        embed_dim: int = 128,
+        distill_net_config: Optional[ModelConfigDict] = None,
+        lr: float = 1e-3,
+        alpha: float = 0.5,
+        beta: float = 0.0,
+        intrinsic_reward_coeff: float = 5e-3,
+        normalize: bool = True,
+        random_timesteps: int = 10000,
+        sub_exploration: Optional[FromConfigSpec] = None,
+        **kwargs,
+    ):
+        """Initializes a NovelD exploration scheme.
+
+        Args:
+            action_space: The action space of the environment.
+            framework: The framework used to train the model.
+            model: The model used to train the model.
```

Comment:
Could you complete this args-section? Missing the most useful stuff (how to configure the NovelD algo) :)